timemachine/fe/rbfe.py

```diff
@@ -125,34 +88,81 @@ def __init__(self, omm_system, conf, box):
         self.box = box
 
 
-def generate_samples(st, host_config, temperature, lambda_schedule, n_frames, seed):
+def sample(initial_state, protocol):
+    """
+    Generate a trajectory given an initial state and a simulation protocol
     """
-    Generate samples given a lambda_schedule. Let L be the number
-    of lambda windows T be the number of frames.
 
-    Parameters
-    ----------
-    st: SingleTopology
-        Topology container for the alchemical ligand
+    bound_impls = [p.bound_impl(np.float32) for p in initial_state.U_fns]
+    intg_impl = initial_state.integrator.impl()
+    baro_impl = initial_state.barostat.impl(bound_impls)
 
-    host_config: HostConfig
-        Configuration of the host
+    ctxt = custom_ops.Context(initial_state.x0, initial_state.v0, initial_state.box0, intg_impl, bound_impls, baro_impl)
 
-    temperature: float
-        Temperature of the system, in Kelvins
+    # burn-in
+    ctxt.multiple_steps_U(
+        lamb=initial_state.lamb,
+        n_steps=protocol.burn_in,
+        lambda_windows=[initial_state.lamb],
+        store_u_interval=0,
+        store_x_interval=0,
+    )
 
-    lambda_schedule: np.array(L,)
-        Sequence of lambda windows
+    # a crude, and probably not great, guess on the decorrelation time
+    n_steps = protocol.n_frames * protocol.steps_per_frame
+    all_nrgs, all_coords, all_boxes = ctxt.multiple_steps_U(
+        lamb=initial_state.lamb,
+        n_steps=n_steps,
+        lambda_windows=[initial_state.lamb],
+        store_u_interval=protocol.steps_per_frame,
+        store_x_interval=protocol.steps_per_frame,
+    )
+
+    assert all_coords.shape[0] == protocol.n_frames
+    assert all_boxes.shape[0] == protocol.n_frames
 
+    return all_coords, all_boxes
+
+
+@dataclass
+class SimulationProtocol:
     n_frames: int
-        Number of samples to generate per window
+    burn_in: int
+    steps_per_frame: int
 
-    Returns
-    -------
-    3-tuple of all_frames, all_boxes, all_U_fns
-        all_frames has shape (L,T,N,3), all_boxes has shape (L,T,3,3), all_U_fns has shape (L,)
 
+@dataclass
+class InitialState:
     """
+    An initial contains everything that is needed to bitwise reproduce a trajectory given a SimulationProtocol
+
+    This object can be pickled safely.
+    """
+
+    U_fns: List[CustomOpWrapper]
```

Comment:
naming nit: the name `U_fns` suggests that these are functions rather than potential objects. Would it be preferable to call this `potentials` instead?

Reply:
I agree - note: I should get rename the `HostGuestSystem::get_U_fns()` to `get_potentials` instead

Reply:
(will do the HGS one in deboggling week)